File: Niel/progamming/H24/in1000/innleveringer/innlevering 8/rutenett.py
from random import randint
from celle import Celle

class Rutenett:
    # Konstruktøren med attributter rader, kolonner og rutenett.
    def __init__(self, rader, kolonner):
        self._ant_rader = rader
        self._ant_kolonner = kolonner
        self._rutenett = []
        self._lag_tomt_rutenett()
        
    # Lager et tomt rutenett med gitt antall rader.
    def _lag_tomt_rutenett(self):
        for _ in range(self._ant_rader):
            tom_rad = self._lag_tom_rad()
            self._rutenett.append(tom_rad)

    # Lager en tom rad med None-verdier.
    def _lag_tom_rad(self):
        tom_rad = [None] * self._ant_kolonner
        return tom_rad

    # Fyller rutenettet med tilfeldige celler som enten kan være levende eller døde.
    def fyll_med_tilfeldige_celler(self):
        for rad in range(self._ant_rader):
            for kol in range(self._ant_kolonner):
                self.lag_celle(rad, kol)

    # Lager en celle på gitt rad og kolonne, setter tilfeldig status som levende.
    def lag_celle(self, rad, kol):
        celle = Celle()
        if randint(1,3) == 1:
            celle.sett_levende()
        self._rutenett[rad][kol] = celle

    # Returnerer cellen på gitt rad og kolonne, eller None hvis det er utenfor rutenettet.
    def hent_celle(self, rad, kol):
        if rad >= 0 and kol >= 0:
            if rad < self._ant_rader and kol < self._ant_kolonner:
                return self._rutenett[rad][kol]
            else:
                return None
        else:
            return None

    # Tegner rutenettet til terminalen.
    def tegn_rutenett(self):
        # Tøm terminalen med linjeskift
        for _ in range(3):
            print()
        # Gå gjennom hver rad og kolonne i rutenettet
        for rad in self._rutenett:
            for celle in rad:
                if celle._status == "doed":
                    tegn = "."
                elif celle._status == "levende":
                    tegn = "O"
                print(tegn, end="")
            print()  # Linjeskift etter hver rad

    # Setter naboer for en gitt celle på en gitt posisjon.
    def _sett_naboer(self, rad, kol):
        koordinater = [
            (rad-1, kol),
            (rad-1, kol-1), # venstre oppe
            (rad, kol-1),
            (rad+1, kol-1), #venstre nede
            (rad+1, kol),
            (rad+1, kol+1), #høyre nede
            (rad, kol+1),
            (rad-1, kol+1)  # høyre oppe
        ]
        
        celle = self.hent_celle(rad, kol)
        for k in koordinater:
            nyCelle = self.hent_celle(k[0], k[1])
            if nyCelle is not None:
                celle.legg_til_nabo(nyCelle)

    # Koble alle celler med sine naboer.
    def koble_celler(self):
        for i in range(len(self._rutenett)):
            for j in range(len(self._rutenett[i])):
                self._sett_naboer(i, j)

    # Hent alle celler i rutenettet som en liste.
    def hent_alle_celler(self):
        liste = []
        for r in self._rutenett:
            liste += r
        return liste

    # Tell antallet levende celler i rutenettet.
    def antall_levende(self):
        antall = 0
        for i in range(len(self._rutenett)):
            for j in range(len(self._rutenett[i])):
                celle = self.hent_celle(i, j)
                if celle.er_levende():
                    antall += 1
        return antall

File: Niel/progamming/V25/in1010/innleveringer/innlevering_2/Rutenett.java
public class Rutenett {

    int antRader; // Antall rader
    int antKolonner; // Antall kolonner
    Celle[][] rutene; // array som representerer cellene i rutenettet.

    // Konstruktør som initialiserer rutenett med gitt antall rader og kolonner.
    public Rutenett(int Rader, int Kolonner){
        antRader = Rader;
        antKolonner = Kolonner;
        rutene = new Celle[Rader][Kolonner];
    }

    // Lager en ny celle på spesifisert posisjon, som er levende med 33.33% sannsynlighet, ellers død.
    public void lagCelle(int rad, int kol){
        Celle nyCelle = new Celle();
        if (Math.random() <= 0.3333){
            nyCelle.settLevende();
        }
        rutene[rad][kol] = nyCelle;
    }

    // Fyller hele rutenettet med celler, der hver celle har en 33.33% sjanse for å være levende.
    public void fyllMedTilfeldigeCeller(){
        for (int i = 0; i < antRader; i++){
            for (int j = 0; j < antKolonner; j++) {
                lagCelle(i, j);
            }
        }
    }

    // Returnerer cellen på gitt rad og kolonne, eller null hvis koordinatene er utenfor grensene.
    public Celle hentCelle(int rad, int kolonne){
        if (antRader > rad && 0 <= rad){
            if (antKolonner > kolonne && 0 <= kolonne){
                return (rutene[rad][kolonne]);
            }
        }
        return null;
    }

    // Tegner rutenettet ved å printe statusen til hver celle (levende eller død).
    public void tegnRutenett() {
        for (int i = 0; i < 2; i++) {
            System.out.println();
        }
        
        for (int rx = 0; rx < antRader; rx++) {
            for (int kx = 0; kx < antKolonner; kx++) {
                Celle celle = rutene[rx][kx];
                System.out.print(celle.hentStatusTegn());
            }
            System.out.println();
        }
        
    }

    // Setter naboene til cellen på gitt rad og kolonne.
    public void settNaboer(int rad, int kolonne){
        Celle celle = hentCelle(rad, kolonne);
        for (int i = rad - 1; i <= rad + 1; i++){
            for (int j = kolonne - 1; j <= kolonne + 1; j++) {
                Celle nabo = hentCelle(i, j);
                if (nabo != null && nabo != celle){
                    celle.leggTilNabo(nabo);
                }
            }
        }
    }

    // Knytter alle celler i rutenettet til sine respektive naboer.
    public void kobleAlleCeller() {
        for (int r = 0; r < antRader; r++) {
            for (int k = 0; k < antKolonner; k++) {
                settNaboer(r, k);
            }
        }
    }
    
    // Teller og returnerer antallet levende celler i rutenettet.
    public int antallLevende() {
        int teller = 0;
        for (int r = 0; r < antRader; r++) {
            for (int k = 0; k < antKolonner; k++) {
                Celle celle = hentCelle(r, k);
                if (celle.erLevende()) {
                    teller += 1;
                }
            }
        }
        return teller;
    }
}
